### DISCUSSION TITLE
Facing issue while establishing conn with mssql and python

### DISCUSSION THREAD

**Discussion by MickyMinnie** [sfu-db/connector-x#181] [Q&A]
Facing issue in the connection string. Can anyone help me to solve that?
While establishing connection it is still running.

![image](https://user-images.githubusercontent.com/11172000/143414930-35cc8cd7-832d-4c1a-80a5-961dd975723a.png)

**Comment by wangxiaoying** [comment]
Hi @MickyMinnie , the connection string looks like this: `mssql://{username}:{password}@{hostname}:{port}/{dbname}`. Maybe you might want to remove `tpc:` and only keep `mywarehouse.database.windows.net` for the hostname part. If it still not works, maybe for the password since there is a `@`, you can try to url encode your password like `urllib.parse.quote_plus('admin@123')`.

The entire connection string will be:
```Python
import urllib.parse
f"mssql://admin:{urllib.parse.quote_plus('admin@123')}@mywarehouse.database.windows.net:1433/mywarehouse"
```

**Reply by MickyMinnie** [reply]
No connection was not working. Tried many methods, but didn't work.

**Reply by wangxiaoying** [reply]
Can you show the error message here?

**Comment by MickyMinnie** [comment]
![image](https://user-images.githubusercontent.com/11172000/143686965-15712d39-f3c1-4ea6-b6e5-833b818d8601.png)
Error is "RuntimeError: Timed out in bb8", Can you kindly help me to solve this. This could be a great help for me !!!!

**Reply by wangxiaoying** [reply]
It seems like it time out in establishing connection with the database. Can you make sure that your hostname and port are correct and is reachable from your machine? Maybe you can try to use some client tools such as [mssql-cli](https://github.com/dbcli/mssql-cli) to test connection to the database? `mssql-cli -S <server URL> -d <database name> -U <username> -P <password>`

**Reply by kkartik1** [reply]
Even I am facing issues while trying to connect with MSSQL Server. I have already verified all the connection details through MS SQL Server Management Studio.
>>> import urllib.parse
>>> import connectorx as cx
>>> mssql_url = f"mssql://{urllib.parse.quote_plus('User ID')}:{urllib.parse.quote_plus('Password')}@Server URL:1433/Database"
>>> query = "SELECT * FROM table"
>>> df = cx.read_sql(mssql_url, query)
[2022-01-27T12:02:13Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:14Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:14Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:16Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:19Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:26Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
[2022-01-27T12:02:38Z ERROR tiberius::tds::stream::token] message=Login failed for user 'User ID'. code=18456 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/path/xxxxxxxx/dev/lib/python3.8/site-packages/connectorx/__init__.py", line 118, in read_sql
    result = _read_sql(
RuntimeError: Timed out in bb8

**Reply by kkartik1** [reply]
I have updated server URL but still getting time out error. Although Login failed error gone.
>>> df = cx.read_sql(mssql_url, query)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/pi_amil/kkartik1/dev/lib/python3.8/site-packages/connectorx/__init__.py", line 118, in read_sql
    result = _read_sql(
RuntimeError: Timed out in bb8

**Reply by wangxiaoying** [reply]
Hi @kkartik1 , MS SQL Server Management Studio may use different authentication / connection method to access mssql. Can you try `mssql-cli` mentioned above to see whether it can connect to your database?

**Reply by larrypham** [reply]
I can connect to mssql database by using mssql-cli and similar "mssql:///" connection string.. but using cx.read_sql still be raised same issue as above ("time out in bb8")

**Reply by FZ-next** [reply]
Hey,
Is there fix to this issue?

I am getting the same error message as described by larrypham and MickyMinnie. I can connect via mssql-cli and other means, too, but I am unable to connect to my AzureDB server as MickyMinnie with the address .database.windows.net. The url I define is identical to the one of MickyMinnie and I, too, tried different definitions. For example with tcp and without, with encoding and without etc.

A side note: When I tried out connectorx with a dockerized mssql server on my local machine, everything worked fine and I was able to use it. So I believe it must be something related maybe to AzureDB.

My url looks like
mssql://<uid>:<pwd>@server.database.windows.net:1433/Azuredbname?Encrypt=yes?TrustServerCertificate=yes
As said above, with a dockerized mssql server this works perfect (of course with a different server name and port number)

It would be great if someone can help out and suggest a solution.

**Reply by wangxiaoying** [reply]
Hi @FZ-next , thanks for the detailed info. Can you let me know your configuration of the Azure database so I can reproduce the error? I tried to created an Azure SQL database previously and it could be successfully connected using connectorx. So I'm thinking maybe it's the configuration that makes the difference.

**Reply by FZ-next** [reply]
Hey,
thank you for the quick replay. I can provide you with this txt file. Is this enough?
Best,
[sql.txt](https://github.com/sfu-db/connector-x/files/8124377/sql.txt)

**Reply by wangxiaoying** [reply]
Thank you! I will try to reproduce the error.

**Reply by wangxiaoying** [reply]
Hi @FZ-next , it seems like your azure sql database can only be accessed through Azure Active Directory. Can you try to use connection string follow this format: `mssql://host:port/db?trusted_connection=true` to enable windows authentication and see whether it works? I cannot test it locally for now since I'm not sure how to setup AzureAD under my school account (since I'm not the admin).

**Reply by FZ-next** [reply]
Hey,
thanks for the response. I already have tried that before. No difference.
Yes, you are correct in default only AzureAD is allowed. However, we have configurated it to allow sql authentication. The SQL Authentiaction  works fine on all channels. For example we use it in combination with pyodbc to connect to the AzureDB and execture queries. Here is no issue.

**Reply by wangxiaoying** [reply]
> 

Hi @FZ-next , just to double check if you use the same `trusted_connection=true` parameter instead of `Encrypt=yes?TrustServerCertificate=yes` you mentioned above. Since only `trusted_connection` trigger the windows authentication in connectorx.

Also, it's weird that it could not work on sql authentication since I've successfully tested it previously. Is it possible if you could print the log, like this example:
```Python
import os
os.environ['RUST_LOG']="connectorx=debug,connectorx_python=debug"
import connectorx as cx

url = "mssql://host:port/db?trusted_connection=true"
df = cx.read_sql(url, query)
```

**Reply by matthieugross-bib** [reply]
Hi @wangxiaoying 
As discussed [here](https://github.com/sfu-db/connector-x/discussions/61), I do have the same issue (using version 1.21.5).
I've tried the trusted connection parameter but still have the same error (Timed out in bb8).
On my side, pymssql works fine but I wanted to benefit from the performance of connector-x.
Let me know if you want me to do any tests.
Matt

**Reply by wangxiaoying** [reply]
Hi @matthieugross-bib , thank you! We will try to reproduce the error on our side. Will let you know if we need help.

**Reply by wangxiaoying** [reply]
Hi @matthieugross-bib , sorry for the late response. We think the issue may related to the encryption. Can you install the newest alpha version `0.2.6.a3` and add `encrypt=true` as parameter to the connection url? For example: `mssql://username:password@server:port/database?encrypt=true`

**Reply by fischja** [reply]
Hi @wangxiaoying, I have the same issue when trying to connect to a local mssql db with windows auth (`RuntimeError: Timed out in bb8`). I have tried `mssql://instance/db?trusted_connection=true&encrypt=true` as a connection string along with various combinations of including/excluding the `trusted_connection` and `encrypt` params. Any ideas?

**Reply by wangxiaoying** [reply]
Hi @fischja  Can you make sure that your hostname and port are correct and is reachable from your machine? Maybe you can try to use some client tools such as [mssql-cli](https://github.com/dbcli/mssql-cli) to test connection to the database or `sqlcmd`? If that works, can you try to connect with username and password using `connectorx`? It will help us to narrow down the problem.

**Reply by fischja** [reply]
Hi @wangxiaoying, the following works: `mssql-cli --server MSI --username test --password 1234 --database TestDb` from cmd, while the following connection string does not work from connector-x: `mssql://test:1234@MSI:1433/TestDb` (with/without port, with/without `encrypt=true`)

**Reply by fischja** [reply]
@wangxiaoying Update: I have managed to solve the issue by enabling TCP/IP in Sql Server Configuration Manager, hope this helps someone 🚀.
![image](https://user-images.githubusercontent.com/23638305/169081409-d8c3b5c8-15c7-4584-9160-2cb27a11da15.png)
